### ISSUE TITLE
Exception occurred: The request failed: Google returned a response with code 429. 

### ISSUE THREAD

**Issue by sundios** [GeneralMills/pytrends#561]
Im getting the following error:

Exception occurred: The request failed: Google returned a response with code 429.

I think this could be because Google has a new trends website?

 https://searchengineland.com/google-launches-new-google-trends-portal-394026

Is there a way to fix this issue? Thanks in advance

**Comment by maxwnewcomer** [comment]
Playing around with postman it seems like they just removed the `/api` route. Now something that would look like `/trends/api/explore` seems to work with `/trends/explore`. Looking into non-explore api calls atm. Hopefully it's just as easy as modifying those routes.

**Comment by Jinoy-Varghese** [comment]
Even I'm also facing the same issue from 08/03/2023 : 5:00PM

**Comment by DarkSand** [comment]
Sam issue, who can tell me how to fix it?

**Comment by vimal-quilt** [comment]
Same Issue. when can this be fixed?

**Comment by aliamourad** [comment]
Pretty sure this has to do with the new Google Trends UI. This may take a while to fix.

**Comment by DarkSand** [comment]
It seems to use grecapcha , for prevent crawling data...
It's hard to resolve this problem

**Comment by Syndorik** [comment]
So I guess no easy work around this time?

**Comment by Syndorik** [comment]
Ok I might have found a work around. It appears that now, the first connection to google trends returns a 429, so if you setup the object with :
```pytrend = TrendReq(retries=3)```

It should work. I've tested it on my side and no more 429

**Comment by netvids** [comment]
I am still getting a 429 with retries=3.     raise RetryError(e, request=request)
requests.exceptions.RetryError: HTTPSConnectionPool(host='trends.google.com', port=443): Max retries exceeded with url: /trends/api/explore?hl=en-....Caused by ResponseError('too many 429 error responses'))

**Comment by Jinoy-Varghese** [comment]
[Syndorik](https://github.com/Syndorik) Brooo.... You are Insane. Really thankful to you man. I had to submit a project tomorrow based on this and now everything is working properly.

**Comment by nicktba** [comment]
> 

Wait, this worked for you? I'm still getting 429.

Can you show your code?

**Comment by Jinoy-Varghese** [comment]
pt = TrendReq(retries=3)
pt.build_payload(terms)
df = pt.interest_over_time()

This works

**Comment by maxwnewcomer** [comment]
> Ok I might have found a work around. It appears that now, the first connection to google trends returns a 429, so if you setup the object with : `pytrend = TrendReq(retries=3)`
> 
> It should work. I've tested it on my side and no more 429

This didn't work for me.

 What I did find working (unsurprisingly, yet still possibly useful for some), is replacing the headers in the `request_args` with a valid cookie from browser. 

For those who need it working "now" like @Jinoy-Varghese and the retries dosn't work, try inspecting element on the trends page, go to network, click on a `?geo` request and copy the cookie into your construction of `TrendReq`. 

Should look like: 
```python
p = TrendReq(request_args = {'headers': {'Cookie': NID COOKIE HERE}})
```

**Comment by nicktba** [comment]
> > Ok I might have found a work around. It appears that now, the first connection to google trends returns a 429, so if you setup the object with : `pytrend = TrendReq(retries=3)`
> > It should work. I've tested it on my side and no more 429
> 
> This didn't work for me.
> 
> What I did find working (unsurprisingly, yet still possibly useful for some), is replacing the headers in the `request_args` with a valid cookie from browser.
> 
> For those who need it working "now" like @Jinoy-Varghese and the retries dosn't work, try inspecting element on the trends page, go to network, click on a `?geo` request and copy the cookie into your construction of `TrendReq`.
> 
> Should look like:
> 
> ```python
> p = TrendReq(request_args = {'headers': {'Cookie': NID COOKIE HERE}})
> ```

Isn't there a risk of running into issues doing this?

borrowing cookies is not my preferred method

**Comment by maxwnewcomer** [comment]
@nicktba Yeah definitely not a long term solution. Just was saying in case someone needed a solution rn for a school project like Jinoy. Especially when there isn't a fix implemented yet.

**Comment by nicktba** [comment]
@maxwnewcomer Can you try implementing cURL impersonate into your payload?

The guys over @ SERPAPI have been using it to resolve their issues

**Comment by maxwnewcomer** [comment]
Thanks to @nicktba I have the curl impersonate working (no retries and no cookies needed). Some of the curl_cffi session methods are different than the normal requests module, so will do some updating to the _get_data() method and hopefully push a fix soon. 

Side effect of this push will be including a new required package.

**Comment by nicktba** [comment]
> Thanks to @nicktba I have the curl impersonate working (no retries and no cookies needed). Some of the curl_cffi session methods are different than the normal requests module, so will do some updating to the _get_data() method and hopefully push a fix soon.
> 
> New functionality will be ability to impersonate:
> 
> ```
> chrome99
> chrome100
> chrome101
> chrome104
> chrome107
> chrome110
> chrome99_android
> edge99
> edge101
> safari15_3
> safari15_5
> ```
> 
> Side effect of this push will be including a new required package, curl_cffi.

Amazing! Thanks Max!

Im going to send you an email, lets chat!

**Comment by maxwnewcomer** [comment]
I'm a little confused by the changes google made. It seems like they wanted to make it harder to scrape their trends data, but when you look at the response from browser and from the impersonate enabled version of pytrends, the user type on browser is `USER_TYPE_LEGIT_USER` and the response from the working pytrends is `USER_TYPE_SCRAPER`. This indicates that they know it's scraping, but don't care? Yet, it still breaks the pytrend scraping. Kind of odd.

**Comment by nicktba** [comment]
> I'm a little confused by the changes google made. It seems like they wanted to make it harder to scrape their trends data, but when you look at the response from browser and from the impersonate enabled version of pytrends, the user type on browser is `USER_TYPE_LEGIT_USER` and the response from the working pytrends is `USER_TYPE_SCRAPER`. This indicates that they know it's scraping, but don't care? Yet, it still breaks the pytrend scraping. Kind of odd.

I think, for now at least, they are just trying to categorize those who are scrapers and those who are not. 

It's likely in the near future they will use this to forecast and integrate an API credit system or block scraping overall.

The user-type update was launched earlier this year and disrupted a wide range of unofficial APIs. PyTrends included.

**Comment by maxwnewcomer** [comment]
Funny... I wonder how many paid/open-sourse SEO and SERP tools they broke and will break in the upcoming years.

**Comment by maxwnewcomer** [comment]
Also update on the fix... I could push a fix without retries working rn, but would like to get that figured out first (along with tests). Worst case if I don't hear from the curl_cffi community in a bit I will just add an "impersonate" flag people can add to the `TrendReq` constructor that will flip functionality from the normal `request` module to cURL Impersonate.

**Comment by nicktba** [comment]
> Also update on the fix... I could push a fix without retries working rn, but would like to get that figured out first (along with tests). Worst case if I don't hear from the curl_cffi community in a bit I will just add an "impersonate" flag people can add to the `TrendReq` constructor that will flip functionality from the normal `request` module to cURL Impersonate.

Awesome! Ill patiently wait for that update.

Thanks for your effort

**Comment by maxwnewcomer** [comment]
Just opened that PR, should work for basic usage. No testing, retries, or confirmed proxy usage with that code. still wip.

**Comment by jesvinc** [comment]
I believe I found a much simpler solution to @maxwnewcomer's. The request made in `GetGoogleCookie` is a `GET`, but it responds with a empty `200` response. But if you instead make do a `POST`, the API correctly responds with a cookie. The fix is simply to change [this line ](https://github.com/GeneralMills/pytrends/blob/master/pytrends/request.py#L88) to be `requests.post`.

**Comment by maxwnewcomer** [comment]
Hahahah sick!! Funny how it can be that simple. I do however think the cURL impersonate functionality might be nice to have in the future. I can add that change to my PR or create your own, up to you!

**Comment by jesvinc** [comment]
Haha yeah, I was shocked to find that out. Since it's so simple, I'm fine with you adding that change to your PR. I'll just add a code comment to it.

**Comment by AbhishekThalanki** [comment]
Tried @MartinNowak's patch but the issue still persists. I've got 50 keywords out of which it returns data for 1 and then throws the error, 
`requests.exceptions.RetryError: HTTPSConnectionPool(host='trends.google.com', port=443): Max retries exceeded with url: /trends/api/explore?hl=en-GB&tz=360&req=%7B%22comparisonItem%22%3A+%5B%7B%22keyword%22%3A+%22accessorize%22%2C+%
22time%22%3A+%222018-07-01+2023-03-12%22%2C+%22geo%22%3A+%22US%22%7D%5D%2C+%22category%22%3A+0%2C+%22property%22%3A+%22%22%7D (Caused by ResponseError('too many 429 error responses'))
`

**Comment by arthii17** [comment]
I am also facing the issue since last 4 days

**Comment by maxwnewcomer** [comment]
@AbhishekThalanki and @arthii17, feel free to pull my fork on the pull request #563. The impersonate feature has seemed to work for me.

**Comment by danielfree** [comment]
face the same issue, adding all local cookies to TrendReq(requests_args=..) seems to fix it temporarily

**Comment by saipavan0798** [comment]
can someone help me, with where and how to find "NID COOKIE", please?

**Comment by gilbertovilarunc** [comment]
Same issue here.

**Comment by gdavoian** [comment]
> I believe I found a much simpler solution to @maxwnewcomer's. The request made in `GetGoogleCookie` is a `GET`, but it responds with a empty `200` response. But if you instead make do a `POST`, the API correctly responds with a cookie. The fix is simply to change [this line ](https://github.com/GeneralMills/pytrends/blob/master/pytrends/request.py#L88) to be `requests.post`.

I would like to confirm that the proposed solution has worked for me perfectly. Thank you @maxwnewcomer!

Please see [my comment](https://github.com/GeneralMills/pytrends/pull/563#issuecomment-1466338941) if you need a temporary workaround to make your code work.

**Comment by karam-khanna** [comment]
Have tried above solutions but seem to still get 429s consistently

**Comment by totencrab** [comment]
This solution posted by @ckosmic [here](https://github.com/GeneralMills/pytrends/pull/563#issuecomment-1464193409) works consistently for me.

In the pytrends request.py file, at line 76 and 89, insert explore before /?geo.

so ` f'{BASE_TRENDS_URL}?geo={self.hl[-2:]}',` becomes so ` f'{BASE_TRENDS_URL}explore/?geo={self.hl[-2:]}',`

**Comment by vikas-sp-97** [comment]
@totencrab Thanks for this bit of information, it was really helpful! 👍

**Comment by aalyousfi** [comment]
Any estimate on when the package will be updated with a fix?
Thanks.

**Comment by danielfree** [comment]
for ppl who want a fix now, you can use a selenium webdriver to visit the webpage once and extract cookie, then add it into TrendReq()

```
def get_cookie():
    options = webdriver.ChromeOptions()
    options.add_argument("--headless")
    driver = webdriver.Chrome(options=options)
    driver.get("https://trends.google.com/")
    time.sleep(5)
    cookie = driver.get_cookie("NID")["value"]
    driver.quit()
    return cookie

nid_cookie = f"NID={get_cookie()}"
pytrends = TrendReq(
      ...
      requests_args={"headers": {"Cookie": nid_cookie}}
)
```

**Comment by bobo17chen** [comment]
Have tried the solution posted by @ckosmic  but seem to still get 429s consistently

**Comment by srgh1367** [comment]
@totencrab Thanks.

**Comment by Michae94** [comment]
> I believe I found a much simpler solution to @maxwnewcomer's. The request made in `GetGoogleCookie` is a `GET`, but it responds with a empty `200` response. But if you instead make do a `POST`, the API correctly responds with a cookie. The fix is simply to change [this line ](https://github.com/GeneralMills/pytrends/blob/master/pytrends/request.py#L88) to be `requests.post`.

I just wanted to thank you for this as it completely solve the issue I had.

**Comment by emlazzarin** [comment]
I believe we resolved this with #570 which is now included in the v4.9.1 release. Thank you all for helping uncover the underlying issues!

**Comment by qcgm1978** [comment]
> > I believe I found a much simpler solution to @maxwnewcomer's. The request made in `GetGoogleCookie` is a `GET`, but it responds with a empty `200` response. But if you instead make do a `POST`, the API correctly responds with a cookie. The fix is simply to change [this line ](https://github.com/GeneralMills/pytrends/blob/master/pytrends/request.py#L88) to be `requests.post`.
> 
> I would like to confirm that the proposed solution has worked for me perfectly. Thank you @maxwnewcomer!
> 
> Please see [my comment](https://github.com/GeneralMills/pytrends/pull/563#issuecomment-1466338941) if you need a temporary workaround to make your code work until the fix has been added to the library.

I encountered this problem again. The version of pytrends is 4.9.2, but changing 'get' to 'post' in GetGoogleCookie function works.

**Comment by alaaddinsarac** [comment]
> I believe I found a much simpler solution to @maxwnewcomer's. The request made in `GetGoogleCookie` is a `GET`, but it responds with a empty `200` response. But if you instead make do a `POST`, the API correctly responds with a cookie. The fix is simply to change [this line ](https://github.com/GeneralMills/pytrends/blob/master/pytrends/request.py#L88) to be `requests.post`.

thanks for the solution. I was going to lose my mind over this

**Comment by camirojasguajardo** [comment]
> I believe I found a much simpler solution to @maxwnewcomer's. The request made in `GetGoogleCookie` is a `GET`, but it responds with a empty `200` response. But if you instead make do a `POST`, the API correctly responds with a cookie. The fix is simply to change [this line ](https://github.com/GeneralMills/pytrends/blob/master/pytrends/request.py#L88) to be `requests.post`.

Hi! does anyone know if this solution still works? im using pytrends 4.9.2

**Comment by AhmadKhanSenpai** [comment]
> > I believe I found a much simpler solution to @maxwnewcomer's. The request made in `GetGoogleCookie` is a `GET`, but it responds with a empty `200` response. But if you instead make do a `POST`, the API correctly responds with a cookie. The fix is simply to change [this line ](https://github.com/GeneralMills/pytrends/blob/master/pytrends/request.py#L88) to be `requests.post`.
> 
> Hi! does anyone know if this solution still works? im using pytrends 4.9.2

no it does not work, I was using it in my project and it was working fine but the next day its gives me the Too many Request error code 429

**Comment by Dojohn2004** [comment]
> > > I believe I found a much simpler solution to @maxwnewcomer's. The request made in `GetGoogleCookie` is a `GET`, but it responds with a empty `200` response. But if you instead make do a `POST`, the API correctly responds with a cookie. The fix is simply to change [this line ](https://github.com/GeneralMills/pytrends/blob/master/pytrends/request.py#L88) to be `requests.post`.
> > 
> > 
> > Hi! does anyone know if this solution still works? im using pytrends 4.9.2
> 
> no it does not work, I was using it in my project and it was working fine but the next day its gives me the Too many Request error code 429

same here, it is not working for me.

**Comment by Dojohn2004** [comment]
> > > > I believe I found a much simpler solution to @maxwnewcomer's. The request made in `GetGoogleCookie` is a `GET`, but it responds with a empty `200` response. But if you instead make do a `POST`, the API correctly responds with a cookie. The fix is simply to change [this line ](https://github.com/GeneralMills/pytrends/blob/master/pytrends/request.py#L88) to be `requests.post`.
> > > 
> > > 
> > > Hi! does anyone know if this solution still works? im using pytrends 4.9.2
> > 
> > 
> > no it does not work, I was using it in my project and it was working fine but the next day its gives me the Too many Request error code 429
> 
> same here, it is not working for me.

I found it works sometimes, it works for few weeks ago, but now is not working again

**Comment by sumitsihag123** [comment]
Hii, Google trend start blocking again. It is observed that this month we are not able to scrape data properly and it start working fine for 3-4 hours but again showing 429 error or malformed error.

**Comment by wheeler88** [comment]
confirm its not working

**Comment by NathanBForBetter** [comment]
Ditto. Not working for me either. Did curl method and nid method.

**Comment by jabeztadesse** [comment]
What is the current official fix?

**Comment by sumitsihag123** [comment]
Hii,

Seems like the google trend again starting working. When it stopped working
then nothing worked. I have tried on multiple platforms. Google trend
detects hitting and blocked the user for some days as per my concern.

**Comment by bdetry** [comment]
Thanks @maxwnewcomer, working on this date.

**Comment by unytics** [comment]
I also got some 429 issues while building a bigquery function to help people load trends into bigquery. 

The only way I solved my problem was to implement @nicktba suggestion: use cURL impersonate (to be honest it's not perfect but it's way better).

I tried to patch pytrends but for some reasons, it failed. So, I started from scratch (with inspiration from this [stackoverflow answer](https://stackoverflow.com/questions/50571317/pytrends-the-request-failed-google-returned-a-response-with-code-429#answer-77485741) which caught my eye after @nicktba comment)

Here is: 

- my function [documentation](https://unytics.io/bigfunctions/bigfunctions/load_google_trends/)
- my function [python code](https://github.com/unytics/bigfunctions/blob/main/bigfunctions/get_data/get_google_trends.yaml#L132) with curl impersonate. 


Besides, if some people here are on BigQuery, they can use this function for free by simply copy & pasting the function example in the documentation. No install is needed.


(ping @maxwnewcomer for info)
